### PR TITLE
 policy: Add Clang thread safety annotations for variables guarded by cs_feeEstimator

### PR DESCRIPTION
**Pull request by practicalswift**
* Add Clang thread safety annotations for variables guarded by `cs_feeEstimator`
* ~~Add missing `cs_feeEstimator` locks~~

### CODE REVIEW

File: src/policy/fees.h [hunk 1]
```diff
@@ -230,10 +230,12 @@ class CBlockPolicyEstimator
     unsigned int HighestTargetTracked(FeeEstimateHorizon horizon) const;
 
 private:
-    unsigned int nBestSeenHeight;
-    unsigned int firstRecordedHeight;
-    unsigned int historicalFirst;
-    unsigned int historicalBest;
+    mutable CCriticalSection cs_feeEstimator;
```
Comment (MarcoFalke) [inline src/policy/fees.h]: Might as well rename to `m_cs_fee_estimator` according to developer notes.

**Comment by practicalswift** [comment]
@MarcoFalke Good idea. Now rename from `cs_feeEstimator` to `m_cs_fee_estimator`. Please re-review :-)

File: src/policy/fees.cpp [hunk 2]
```diff
@@ -710,6 +711,7 @@ CFeeRate CBlockPolicyEstimator::estimateRawFee(int confTarget, double successThr
 
 unsigned int CBlockPolicyEstimator::HighestTargetTracked(FeeEstimateHorizon horizon) const
 {
+    LOCK(cs_feeEstimator);
```
Comment (Empact) [inline src/policy/fees.cpp]: I don't think we need exclusive locking for these operations, as `GetMaxConfirms` so concurrent access wouldn't result in incoherent data.

File: src/policy/fees.cpp [hunk 3]
```diff
@@ -672,6 +672,8 @@ CFeeRate CBlockPolicyEstimator::estimateFee(int confTarget) const
 
 CFeeRate CBlockPolicyEstimator::estimateRawFee(int confTarget, double successThreshold, FeeEstimateHorizon horizon, EstimationResult* result) const
 {
+    LOCK(cs_feeEstimator);
```
Comment (Empact) [inline src/policy/fees.cpp]: `get` is a read of the ptr value, which is only updated in the constructor, so this lock is unnecessary AFAICT. `PT_GUARDED_BY` could be more appropriate.

File: src/policy/fees.cpp [hunk 4]
```diff
@@ -710,6 +711,7 @@ CFeeRate CBlockPolicyEstimator::estimateRawFee(int confTarget, double successThr
 
 unsigned int CBlockPolicyEstimator::HighestTargetTracked(FeeEstimateHorizon horizon) const
 {
+    LOCK(m_cs_fee_estimator);
```
Comment (Empact) [inline src/policy/fees.cpp]: I don't think we need exclusive locking here, as `GetMaxConfirms` is a read so concurrent access wouldn't result in incoherent data.

File: src/policy/fees.cpp [hunk 5]
```diff
@@ -672,6 +672,8 @@ CFeeRate CBlockPolicyEstimator::estimateFee(int confTarget) const
 
 CFeeRate CBlockPolicyEstimator::estimateRawFee(int confTarget, double successThreshold, FeeEstimateHorizon horizon, EstimationResult* result) const
 {
+    LOCK(m_cs_fee_estimator);
```
Comment (Empact) [inline src/policy/fees.cpp]: `get` is a read of the ptr value, which is only updated in the constructor, so this lock is unnecessary AFAICT. `PT_GUARDED_BY` could be more appropriate.

File: src/policy/fees.cpp (hunk 4 above)
Comment (practicalswift) [reply src/policy/fees.cpp]: @Empact Do you mean that the following annotations are incorrect and should be change or removed?

```
    std::unique_ptr<TxConfirmStats> feeStats GUARDED_BY(m_cs_fee_estimator);
    std::unique_ptr<TxConfirmStats> shortStats GUARDED_BY(m_cs_fee_estimator);
    std::unique_ptr<TxConfirmStats> longStats GUARDED_BY(m_cs_fee_estimator);
```

**Comment by practicalswift** [comment]
@Empact @MarcoFalke Updated. Would you mind re-reviewing? The current version is annotations only, so no change in locking. In other words: this PR changes only compile-time checking and no runtime behaviour. Should hence hopefully be trivial to review :-)

**Comment by ken2812221** [comment]
utACK 1a01e1c120809b19ddc9c9c6ee5d6686a90f55b3

File: src/policy/fees.h [hunk 6]
```diff
@@ -241,34 +243,32 @@ class CBlockPolicyEstimator
     };
 
     // map of txids to information about that transaction
-    std::map<uint256, TxStatsInfo> mapMemPoolTxs;
+    std::map<uint256, TxStatsInfo> mapMemPoolTxs GUARDED_BY(m_cs_fee_estimator);
 
     /** Classes to track historical data on transaction confirmations */
     std::unique_ptr<TxConfirmStats> feeStats;
     std::unique_ptr<TxConfirmStats> shortStats;
     std::unique_ptr<TxConfirmStats> longStats;
```
Comment (Empact) [inline src/policy/fees.h]: I'm curious - did you try marking these `PT_GUARDED_BY`? Seems it would be appropriate, based on the lock in `estimateRawFee`.

**Pull request merged**